src/hotspot/cpu/riscv/assembler_riscv.hpp

```diff
@@ -928,209 +890,408 @@ enum operand_size { int8, int16, int32, uint32, int64 };
+    guarantee(is_uimm2(Fmt), "FMT is out of validity");
+    guarantee(is_uimm5(funct5), "Funct5 is out of validity");
+    uint32_t insn = 0;
+    patch((address)&insn,   6, 0, OP_FP_MAJOR);
+    patch((address)&insn, 11,  7, Rd);
+    patch((address)&insn, 14, 12, rm);
+    patch((address)&insn, 19, 15, Rs1);
+    patch((address)&insn, 24, 20, Rs2);
+    patch((address)&insn, 26, 25, Fmt);
+    patch((address)&insn, 31, 27, funct5);
+    emit(insn);
   }
 
-  INSN(feq_s,    0b1010011, 0b010, 0b1010000);
-  INSN(flt_s,    0b1010011, 0b001, 0b1010000);
-  INSN(fle_s,    0b1010011, 0b000, 0b1010000);
-  INSN(feq_d,    0b1010011, 0b010, 0b1010001);
-  INSN(fle_d,    0b1010011, 0b000, 0b1010001);
-  INSN(flt_d,    0b1010011, 0b001, 0b1010001);
-#undef INSN
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(FloatRegister Rd, FloatRegister Rs1, FloatRegister Rs2, RoundingMode rm) {
+    fp_base<Fmt, funct5>(Rd->raw_encoding(), Rs1->raw_encoding(), Rs2->raw_encoding(), rm);
+  }
 
-// Float and Double Arith Instruction
-#define INSN(NAME, op, funct7)                                                                  \
-  void NAME(FloatRegister Rd, FloatRegister Rs1, FloatRegister Rs2, RoundingMode rm = rne) {    \
-    unsigned insn = 0;                                                                          \
-    patch((address)&insn, 6, 0, op);                                                            \
-    patch((address)&insn, 14, 12, rm);                                                          \
-    patch((address)&insn, 31, 25, funct7);                                                      \
-    patch_reg((address)&insn, 7, Rd);                                                           \
-    patch_reg((address)&insn, 15, Rs1);                                                         \
-    patch_reg((address)&insn, 20, Rs2);                                                         \
-    emit(insn);                                                                                 \
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(FloatRegister Rd, FloatRegister Rs1, FloatRegister Rs2, int8_t rm) {
+    fp_base<Fmt, funct5>(Rd->raw_encoding(), Rs1->raw_encoding(), Rs2->raw_encoding(), (RoundingMode)rm);
   }
 
-  INSN(fadd_s,   0b1010011, 0b0000000);
-  INSN(fsub_s,   0b1010011, 0b0000100);
-  INSN(fmul_s,   0b1010011, 0b0001000);
-  INSN(fdiv_s,   0b1010011, 0b0001100);
-  INSN(fadd_d,   0b1010011, 0b0000001);
-  INSN(fsub_d,   0b1010011, 0b0000101);
-  INSN(fmul_d,   0b1010011, 0b0001001);
-  INSN(fdiv_d,   0b1010011, 0b0001101);
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(Register Rd, FloatRegister Rs1, FloatRegister Rs2, int8_t rm) {
+    fp_base<Fmt, funct5>(Rd->raw_encoding(), Rs1->raw_encoding(), Rs2->raw_encoding(), (RoundingMode)rm);
+  }
 
-#undef INSN
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(FloatRegister Rd, FloatRegister Rs1, int Rs2, int8_t rm) {
```

Comment:
Minor question: Should `int Rs2` be `uint8_t Rs2` for consistency with friends?

Reply:
Yes, thanks.